### PR TITLE
Clarify behavior around namespaces paths

### PR DESCRIPTION
**Pull request by mrunalp**
See #155 
Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

**Comment by wking** [comment]
Instead of “ignore any other config when a namespace path is set”, I'd
rather just error-out if some incompatible values are set [1] and
support the join-and-tweak (or tweak-and-join) where possible [2].
Silently ignoring portions of the config seems like it would be a
source of confusion.

[1]: https://github.com/opencontainers/specs/issues/155#issuecomment-138350425
[2]: https://github.com/opencontainers/specs/issues/155#issuecomment-138099758

**Comment by gao-feng** [comment]
@wking Can not join-and-tweat be implemented by hook, Ignore configuation may confuse someone, how about warning or error-out.

**Comment by laijs** [comment]
The clarify is definitely necessary and important! Thanks!

I think we should also point out which config are the "other config related to that namespace" clearly.
And the implementation should output warning or error if it is hit.

```
Spec.Root
Spec.Mount
LinuxSpec.Linux.RootfsPropagation
RuntimeSpec.Mounts
LinuxRuntime.Devices
...
```

I think this list hasn't contains all

Any thought?

File: runtime-config-linux.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ For more information, see [the man page](http://man7.org/linux/man-pages/man7/na
 Namespaces are specified in the spec as an array of entries.
 Each entry has a type field with possible values described below and an optional path element.
 If a path is specified, that particular file is used to join that type of namespace.
+Also, when a path is specified, a runtime MUST assume that the setup for that particular namespace has already been done and ignore any other config related to that namespace.
```
Comment (philips) [inline runtime-config-linux.md]: Hrm, if there is a join and a create in the runtime I would say it should error, not ignore.

Comment (vishh) [inline runtime-config-linux.md]: +1 for validation and raising an error. We essentially need something like [proto oneof](https://developers.google.com/protocol-buffers/docs/proto3?hl=en#oneof) to get around this issue.

Comment (dqminh) [inline runtime-config-linux.md]: Right. I was wondering if we should describe config and state in a richer language than go struct ( like protobuf ), which we can then serialize to/from json.

**Comment by wking** [comment]
On Wed, Sep 09, 2015 at 12:52:40AM -0700, Gao feng wrote:
> @wking Can not join-and-tweat be implemented by hook…

Lots of existing config stuff *can* be implemented via hooks
(e.g. everything under ‘mounts’).  But we have spec entries for it,
and there's nothing non-sensical about “Join the mnt namespace at
foo/bar and mount a tmpfs at /a/b/c”.  If we thought the “mount a
tmpfs at /a/b/c” was common enough to be in the spec for container
creation, I see no reason to have it *not* be in the spec for
namespace joinin.

> Ignore configuation may confuse someone, how about warning or
> error-out.

Yeah, I prefer explicitly failing to silently ignoring non-sensical
content [1].  I just don't think all “joining an existing namespace
and manipulating that namespace” specs are non-sensical.  But in
today's call, the consensus seemed to be punting on cases like that
until someone tells us a convincing story about why they would be
useful.  Until then, I'm fine erroring out whenever a config specifies
a namespace path and also modifications for that namespace.

[1]: https://github.com/opencontainers/specs/pull/158#issuecomment-138687129

File: runtime-config-linux.md (hunk 1 above)
Comment (mrunalp) [inline runtime-config-linux.md]: @vishh Yes, that is the requirement.
@philips Updated.

File: runtime-config-linux.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ For more information, see [the man page](http://man7.org/linux/man-pages/man7/na
 Namespaces are specified in the spec as an array of entries.
 Each entry has a type field with possible values described below and an optional path element.
 If a path is specified, that particular file is used to join that type of namespace.
+Also, when a path is specified, a runtime MUST assume that the setup for that particular namespace has already been done and error out if there is any other config related to that namespace specified.
```
Comment (wking) [inline runtime-config-linux.md]: “namespace specified” reads a little awkwardly to me.  Maybe “if the config specifies anything else related to that namespace”?

**Comment by gao-feng** [comment]
@wking 
`Yeah, I prefer explicitly failing to silently ignoring non-sensical
content [1].  I just don't think all “joining an existing namespace
and manipulating that namespace” specs are non-sensical.  But in
today's call, the consensus seemed to be punting on cases like that
until someone tells us a convincing story about why they would be
useful.  Until then, I'm fine erroring out whenever a config specifies
a namespace path and also modifications for that namespace.
`
Thanks for your information, 
Let's disallow this behavior to get rid of the possible mistake.

**Comment by wking** [comment]
02b456e9 looks good to me.

**Comment by gao-feng** [comment]
LGTM

**Comment by vbatts** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**